### PR TITLE
hotfix ETP-3471: Add stress tests

### PR DESCRIPTION
**Pull request by Emi-Polliotti**
Includes stress tests for orders, purchase orders, invoices, goods receipts, goods shipments, queries, end-to-end and concurrent scenarios.

### CODE REVIEW

File: src-test/src/org/openbravo/test/stress/InvoiceStressTest.java [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+public class InvoiceStressTest extends StressTestBase {
+
+  private static final String STRESS_DOC_PREFIX = "STRESS-INV-";
+  private static final String LOG_TOTAL_TIME = "[STRESS-INV] Total time: ";
+  private static final String SO_UNIT_PRICE_STR = "25.00";
+  private static final String PO_UNIT_PRICE_STR = "9.08";
+  private static final BigDecimal UNIT_PRICE = new BigDecimal(SO_UNIT_PRICE_STR);
+
+  private static final long MAX_MS_PER_INVOICE_CREATION = 15;
+  private static final long MAX_MS_PER_INVOICE_COMPLETE = 150;
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/InvoiceStressTest.java:27]: **Suggestion**

Consider making performance thresholds configurable via system properties or environment variables to allow adjustments based on the execution environment performance.

```java
private static final long MAX_MS_PER_INVOICE_CREATION = Long.parseLong(System.getProperty("stress.test.threshold.creation", "15"));
```

**Rationale**: Hardcoded timing thresholds in stress tests can be brittle and cause intermittent failures in different environments (e.g., CI/CD pipelines vs. local developer machines).

File: src-test/src/org/openbravo/test/stress/InvoiceStressTest.java [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+  // ==========================================================================
+
+  @Test
+  public void testBulkInvoiceQuery_Performance() {
+    int totalInvoices = 200;
+    int linesPerInvoice = 3;
+    runBulkInvoiceCreation(totalInvoices, linesPerInvoice, true);
+
+    OBDal.getInstance().getSession().clear();
+
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/InvoiceStressTest.java:126]: **Suggestion**

Use the existing 'flushAndClear()' helper method consistently throughout the class instead of calling session.clear() directly, even if a flush was performed recently.

```java
flushAndClear();
```

**Rationale**: Using a consistent helper like flushAndClear() improves readability and ensures that all pending changes are persisted before the session is cleared, preventing data loss in the test context.

File: src-test/src/org/openbravo/test/stress/GoodsReceiptStressTest.java [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        long poCreateStart = System.nanoTime();
+
+        Order po = createOrder(orderCtx.docType, orderCtx.bp, orderCtx.warehouse,
+            orderCtx.priceList, orderCtx.paymentTerm, orderCtx.paymentMethod,
+            orderCtx.currency, STRESS_DOC_PREFIX + "PO-" + i, false);
+        OBDal.getInstance().save(po);
+
+        List<String> orderLineIds = new ArrayList<>();
+        for (int j = 0; j < linesPerDocument; j++) {
+          OrderLine orderLine = createOrderLine(po, orderCtx.product, orderCtx.tax,
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/GoodsReceiptStressTest.java:253]: **Suggestion**

Remove the redundant flush() inside the OrderLine creation loop to improve performance during stress tests.

```java
        for (int j = 0; j < linesPerDocument; j++) {
          OrderLine orderLine = createOrderLine(po, orderCtx.product, orderCtx.tax,
              (long) (j + 1) * 10,
              new BigDecimal(j + 1), new BigDecimal("9.08"), new BigDecimal("9.08"));
          OBDal.getInstance().save(orderLine);
          orderLineIds.add(orderLine.getId());
        }

        OBDal.getInstance().flush();
```

**Rationale**: In Etendo/Openbravo, calling flush() inside a loop when creating multiple lines can cause significant performance overhead due to repeated round-trips to the database and trigger execution for each line. Since UUIDs are generated on save() and a final flush() is called after the loop, the per-line flush is redundant unless specific logic requires it.

File: src-test/src/org/openbravo/test/stress/ConcurrentStressTest.java [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+          String docNo = STRESS_DOC_PREFIX + prefix + "-T" + threadId + "-" + i;
+
+          Order order = createOrder(ctx.docType, ctx.bp, ctx.warehouse, ctx.priceList,
+              ctx.paymentTerm, ctx.paymentMethod, ctx.currency, docNo, isSalesOrder);
+          OBDal.getInstance().save(order);
+
+          for (int j = 0; j < linesPerOrder; j++) {
+            OrderLine line = createOrderLine(order, ctx.product, ctx.tax,
+                (long) (j + 1) * 10, new BigDecimal(j + 1), new BigDecimal(unitPrice),
+                new BigDecimal(unitPrice));
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/ConcurrentStressTest.java:580]: **Suggestion**

Consider using BigDecimal.valueOf() and pre-instantiating the price BigDecimal outside the nested loop.

```java
(long) (j + 1) * 10, BigDecimal.valueOf(j + 1), unitPriceBD,
```

**Rationale**: Instantiating new BigDecimal objects from strings or integers inside high-frequency loops during stress tests can be optimized by using BigDecimal.valueOf() or pre-instantiating the price outside the loop.

File: src-test/src/org/openbravo/test/stress/ConcurrentStressTest.java [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+          }
+
+          SessionHandler.getInstance().commitAndClose();
+        } catch (Exception e) {
+          errors.add(LOG_THREAD + threadId + ": " + e.getClass().getSimpleName()
+              + ": " + e.getMessage());
+          try {
+            OBDal.getInstance().rollbackAndClose();
+          } catch (Exception ex) {
+            // ignore
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/ConcurrentStressTest.java:274]: **Suggestion**

Log the exception during rollback even if the logic proceeds to ignore it.

```java
catch (Exception ex) { log.error(\"Error during rollback cleanup\", ex); }
```

**Rationale**: While it is common to ignore exceptions during a secondary cleanup rollback to prevent masking the primary error, logging the event at an error or debug level can help identify connection pool exhaustion or database locking issues during stress tests.

File: src-test/src/org/openbravo/test/stress/GoodsShipmentStressTest.java [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+          log.info("[STRESS-SHIP] After " + (i + 1) + " shipments - Memory: "
+              + ((memCurrent - memBefore) / (1024 * 1024)) + " MB, Time: " + elapsed + " ms");
+
+          ctx.reload(SHIPMENT_DOCTYPE_ID, BPARTNER_ID,
+              SO_WAREHOUSE_ID, SO_LOCATOR_ID, SO_PRODUCT_ID);
+        }
+      }
+
+      OBDal.getInstance().flush();
+      long totalTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/GoodsShipmentStressTest.java:149]: **Suggestion**

Move the 'OBDal.getInstance().flush()' call outside the OrderLine creation loop in 'runBulkShipmentCreateAndComplete'.

```java
        }
        OBDal.getInstance().flush();
```

**Rationale**: Calling flush() inside a loop for every record (OrderLine) triggers multiple database roundtrips, which negatively impacts the performance of the stress test. Moving the flush() outside the loop ensures all lines are persisted in a single operation before the subsequent stored procedure call, while maintaining the necessary data state.

File: src-test/src/org/openbravo/test/stress/PurchaseOrderStressTest.java [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        OBDal.getInstance().flush();
+        createTime += TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - createStart);
+
+        completeTime += completeOrderAndVerify(order);
+
+        completed++;
+
+        if (completed % 10 == 0) {
+          OBDal.getInstance().getSession().clear();
+          ctx.reload(PO_DOCTYPE_ID, BPARTNER_ID, PO_WAREHOUSE_ID, PO_PRICELIST_ID,
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/PurchaseOrderStressTest.java:255]: **Suggestion**

Use the flushAndClear() helper method instead of calling session().clear() directly for consistency with other methods in the same class.

```java
+          flushAndClear();
```

**Rationale**: Inconsistency in session management methods. While runBulkOrderCreation uses flushAndClear(), this method uses session().clear() directly. Using a consistent helper method improves readability and ensures that pending changes are handled correctly before clearing the session.

File: src-test/src/org/openbravo/test/stress/PurchaseOrderStressTest.java [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.openbravo.dal.core.OBContext;
+import org.openbravo.dal.service.OBDal;
+import org.openbravo.model.common.order.Order;
+import org.openbravo.model.common.order.OrderLine;
+
+/** Stress tests for Purchase Order creation and completion. */
+public class PurchaseOrderStressTest extends StressTestBase {
+
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/PurchaseOrderStressTest.java:17]: **Suggestion**

Add the mandatory Etendo license header at the beginning of the file.

```java
+/*
+ *************************************************************************
+ * The contents of this file are subject to the Etendo License
+ * (the "License"), you may not use this file except in compliance
+ * with the License.
+ * You may obtain a copy of the License at
+ * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
+ * Software distributed under the License is distributed on an
+ * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
+ * implied. See the License for the specific language governing rights
+ * and limitations under the License.
+ * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
+ * All Rights Reserved.
+ * Contributor(s): Futit Services S.L.
+ *************************************************************************
+ */
+package org.openbravo.test.stress;
+...
+public class PurchaseOrderStressTest extends StressTestBase {
```

**Rationale**: The PR title and branch (ETP-3471) indicate this is part of an Etendo Platform task. New Java files should ideally include the standard Etendo license header to ensure compliance and consistency across the codebase. Since the package is org.openbravo, this is treated as a non-blocking suggestion.

File: src-test/src/org/openbravo/test/stress/OrderStressTest.java [hunk 9]
```diff
@@ -0,0 +1,357 @@
+package org.openbravo.test.stress;
+
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/OrderStressTest.java:2]: **Suggestion**

Add the standard Etendo license header at the beginning of the new file.

```java
/*
 *************************************************************************
 * The contents of this file are subject to the Etendo License
 * (the "License"), you may not use this file except in compliance
 * with the License.
 * You may obtain a copy of the License at
 * https://github.com/etendosoftware/etendo_core/blob/main/legal/Etendo_license.txt
 * Software distributed under the License is distributed on an
 * "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing rights
 * and limitations under the License.
 * All portions are Copyright (C) 2021-2024 FUTIT SERVICES, S.L
 * All Rights Reserved.
 * Contributor(s): Futit Services S.L.
 *************************************************************************
 */
package org.openbravo.test.stress;
```

**Rationale**: New source files should include the standard Etendo license header to ensure compliance and properly document copyright.

File: src-test/src/org/openbravo/test/stress/OrderStressTest.java [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+      OBDal.getInstance().flush();
+      long totalTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+
+      log.info("[STRESS] === FLUSH INTERVAL TEST RESULTS ===");
+      log.info(
+          "[STRESS] Total orders created: " + created + " (flush every " + flushInterval + ")");
+      log.info(LOG_TOTAL_TIME + totalTime + " ms");
+      log.info(LOG_AVG_PER_ORDER + (totalTime / created) + " ms");
+
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/OrderStressTest.java:137]: **Suggestion**

Use Math.max(created, 1) to avoid potential division by zero.

```java
      log.info(LOG_AVG_PER_ORDER + (totalTime / Math.max(created, 1)) + " ms");
```

**Rationale**: Using Math.max(created, 1) prevents a potential ArithmeticException (division by zero) if no orders were successfully created, which aligns with the pattern used in other methods of the same class (e.g., line 241).

File: src-test/src/org/openbravo/test/stress/OrderStressTest.java [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+      long memAfter = runtime.totalMemory() - runtime.freeMemory();
+
+      log.info("[STRESS] === MEMORY USAGE TEST RESULTS ===");
+      log.info("[STRESS] Total orders: " + totalOrders + " (5 lines each)");
+      log.info(LOG_TOTAL_TIME + totalTime + " ms");
+      log.info(
+          "[STRESS] Memory delta: " + ((memAfter - memBefore) / (1024 * 1024)) + " MB");
+
+      Assert.assertTrue("Memory usage should be reported", totalTime > 0);
+    } finally {
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/OrderStressTest.java:196]: **Suggestion**

Update the assertion message to accurately describe the condition being checked.

```java
      Assert.assertTrue("Test execution time should be greater than zero", totalTime > 0);
```

**Rationale**: The assertion message 'Memory usage should be reported' is misleading because the condition being checked is 'totalTime > 0'. The message should accurately reflect the failure condition.

File: src-test/src/org/openbravo/test/stress/StressTestBase.java [hunk 12]
```diff
@@ -0,0 +1,752 @@
...
+    order.setDocumentNo(docNo);
+    order.setDocumentStatus("DR");
+    order.setDocumentAction("CO");
+    order.setDocumentType(docType);
+    order.setTransactionDocument(docType);
+    order.setOrderDate(now);
+    order.setAccountingDate(now);
+    order.setBusinessPartner(bp);
+    order.setPartnerAddress(bp.getBusinessPartnerLocationList().get(0));
+    order.setPaymentMethod(paymentMethod);
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/StressTestBase.java:577]: > [!WARNING]
**⚠️ Blocking Issue**

Add a check to verify that the Business Partner has at least one location before accessing the first element.

```java
    if (bp.getBusinessPartnerLocationList().isEmpty()) {
      throw new org.openbravo.base.exception.OBException("Business Partner " + bp.getName() + " has no locations defined.");
    }
    order.setPartnerAddress(bp.getBusinessPartnerLocationList().get(0));
```

**Rationale**: Dereferencing the first element of a list without checking its size can lead to an IndexOutOfBoundsException if the list is empty. This is especially risky even in tests if the environment setup varies.

File: src-test/src/org/openbravo/test/stress/StressTestBase.java [hunk 13]
```diff
@@ -0,0 +1,752 @@
...
+    shipmentReceipt.setDocumentNo(docNo);
+    shipmentReceipt.setDocumentStatus("DR");
+    shipmentReceipt.setDocumentAction("CO");
+    shipmentReceipt.setDocumentType(docType);
+    shipmentReceipt.setMovementType(movementType);
+    shipmentReceipt.setMovementDate(now);
+    shipmentReceipt.setAccountingDate(now);
+    shipmentReceipt.setBusinessPartner(bp);
+    shipmentReceipt.setPartnerAddress(bp.getBusinessPartnerLocationList().get(0));
+    shipmentReceipt.setWarehouse(warehouse);
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/StressTestBase.java:653]: > [!WARNING]
**⚠️ Blocking Issue**

Add a check to verify that the Business Partner has at least one location before accessing the first element.

```java
    if (bp.getBusinessPartnerLocationList().isEmpty()) {
      throw new org.openbravo.base.exception.OBException("Business Partner " + bp.getName() + " has no locations defined.");
    }
    shipmentReceipt.setPartnerAddress(bp.getBusinessPartnerLocationList().get(0));
```

**Rationale**: Dereferencing the first element of a list without checking its size can lead to an IndexOutOfBoundsException if the list is empty.

File: src-test/src/org/openbravo/test/stress/StressTestBase.java [hunk 14]
```diff
@@ -0,0 +1,752 @@
...
+    invoice.setDocumentNo(docNo);
+    invoice.setDocumentStatus("DR");
+    invoice.setDocumentAction("CO");
+    invoice.setDocumentType(docType);
+    invoice.setTransactionDocument(docType);
+    invoice.setInvoiceDate(now);
+    invoice.setAccountingDate(now);
+    invoice.setBusinessPartner(bp);
+    invoice.setPartnerAddress(bp.getBusinessPartnerLocationList().get(0));
+    invoice.setPaymentMethod(paymentMethod);
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/StressTestBase.java:719]: > [!WARNING]
**⚠️ Blocking Issue**

Add a check to verify that the Business Partner has at least one location before accessing the first element.

```java
    if (bp.getBusinessPartnerLocationList().isEmpty()) {
      throw new org.openbravo.base.exception.OBException("Business Partner " + bp.getName() + " has no locations defined.");
    }
    invoice.setPartnerAddress(bp.getBusinessPartnerLocationList().get(0));
```

**Rationale**: Dereferencing the first element of a list without checking its size can lead to an IndexOutOfBoundsException if the list is empty.

File: src-test/src/org/openbravo/test/stress/StressTestBase.java [hunk 15]
```diff
@@ -0,0 +1,752 @@
...
+    try {
+      for (int i = 0; i < triggers.length; i++) {
+        OBDal.getInstance().getSession()
+            .createNativeQuery(
+                "ALTER TABLE " + tables[i] + " ENABLE TRIGGER " + triggers[i])
+            .executeUpdate();
+      }
+      OBDal.getInstance().flush();
+    } catch (Exception e) {
+      log.error("[STRESS] Warning: Could not re-enable audit triggers: " + e.getMessage());
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/StressTestBase.java:203]: > [!WARNING]
**⚠️ Blocking Issue**

Log the caught exception before attempting a rollback to ensure the cause of the cleanup failure is recorded.

```java
    } catch (Exception e) {
      log.error("[STRESS] Error during cleanup: ", e);
      try {
        OBDal.getInstance().rollbackAndClose();
      } catch (Exception ex) {
        log.error("[STRESS] Cleanup rollback error: ", ex);
      }
    } finally {
```

**Rationale**: The original exception 'e' caught in the cleanup block is completely swallowed. It should be logged or rethrown as an OBException to aid in debugging test failures.

Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/StressTestBase.java:203]: **Suggestion**

Pass the exception object to the log.error method to include the full stack trace.

```java
    } catch (Exception e) {
      log.error("[STRESS] Warning: Could not re-enable audit triggers", e);
    }
```

**Rationale**: Logging only the message (e.getMessage()) misses the stack trace, which is crucial for identifying where and why the error occurred. Passing the exception object to the logger provides the full context.

File: src-test/src/org/openbravo/test/stress/QueryStressTest.java [hunk 16]
```diff
@@ -0,0 +1,770 @@
...
+    try {
+      long start = System.nanoTime();
+      Number count = (Number) OBDal.getInstance().getSession()
+          .createNativeQuery("SELECT count(*) FROM c_order")
+          .uniqueResult();
+      long elapsed = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+
+      log.info("[STRESS-QUERY] Order count: " + count + " in " + elapsed + " ms");
+      Assert.assertTrue("Expected orders > 0", count.longValue() > 0);
+      Assert.assertTrue("Count query too slow: " + elapsed + " ms (max: 500 ms)",
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/QueryStressTest.java]: > [!WARNING]
**⚠️ Blocking Issue**

Add a null check or an assertion to ensure the query result is not null before accessing its properties.

```java
      Assert.assertNotNull("Count result should not be null", count);
      Assert.assertTrue("Expected orders > 0", count.longValue() > 0);
```

**Rationale**: The `uniqueResult()` method can return `null` if the query execution fails or in specific edge cases. Dereferencing it with `.longValue()` without a null check will cause a NullPointerException, violating the Etendo safety rules for query results.

File: src-test/src/org/openbravo/test/stress/QueryStressTest.java [hunk 17]
```diff
@@ -0,0 +1,770 @@
...
+        try {
+          OBContext.setOBContext(CLIENT_USER_ID, CLIENT_ROLE_ID, CLIENT_CLIENT_ID, CLIENT_ORG_ID);
+          OBContext.setAdminMode(true);
+
+          for (int i = 0; i < queriesPerThread; i++) {
+            Number count = (Number) OBDal.getInstance().getSession()
+                .createNativeQuery(COUNT_SO_QUERY)
+                .uniqueResult();
+            Assert.assertTrue(count.longValue() > 0);
+            completed++;
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/QueryStressTest.java:636]: > [!WARNING]
**⚠️ Blocking Issue**

Add a null check or an assertion before calling `.longValue()` on the query result.

```java
            Assert.assertNotNull("Count result should not be null", count);
            Assert.assertTrue(count.longValue() > 0);
```

**Rationale**: Same as in the sequential test case, the query result from `uniqueResult()` is accessed without validation inside the concurrent thread, posing a risk of NullPointerException.

File: src-test/src/org/openbravo/test/stress/QueryStressTest.java [hunk 18]
```diff
@@ -0,0 +1,770 @@
...
+  // ==========================================================================
+
+  @Test
+  public void testOrderFilterByDateRange() {
+    OBContext.setAdminMode(true);
+    try {
+      long start = System.nanoTime();
+      OBQuery<Order> query = OBDal.getInstance()
+          .createQuery(Order.class,
+              "orderDate >= '2024-01-01' and orderDate <= '2024-12-31'");
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/QueryStressTest.java:85]: **Suggestion**

Consider using named parameters for date filters instead of hardcoded strings.

```java
          .createQuery(Order.class,
              "orderDate >= :startDate and orderDate <= :endDate")
          .setNamedParameter("startDate", someDate)
          .setNamedParameter("endDate", someDate2);
```

**Rationale**: Using hardcoded date strings in HQL queries can lead to parsing errors or inconsistent behavior across different database configurations or locale settings. It is a best practice to use parameterized queries with `Date` objects.

File: src-test/src/org/openbravo/test/stress/QueryStressTest.java [hunk 19]
```diff
@@ -0,0 +1,770 @@
...
+            }
+          }
+
+          SessionHandler.getInstance().commitAndClose();
+        } catch (Exception e) {
+          errors.add("Thread " + threadId + ": " + e.getMessage());
+          try {
+            OBDal.getInstance().rollbackAndClose();
+          } catch (Exception ex) {
+            // ignore
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/QueryStressTest.java:649]: **Suggestion**

Consider adding a debug log or a comment explaining why the exception is being ignored.

```java
          } catch (Exception ex) {
            log.debug("Error during rollback", ex);
          }
```

**Rationale**: Empty catch blocks are generally discouraged in the Etendo framework as they can swallow important errors. While ignoring an exception during rollback in a failure handler is common, it is better to at least log it at the debug level.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/org/openbravo/test/stress/QueryStressTest.java [hunk 20]
```diff
@@ -0,0 +1,770 @@
...
+
+        OBDal.getInstance().getSession().clear();
+      }
+
+      log.info("[STRESS-QUERY] Pagination (" + totalPages + " pages x "
+          + pageSize + "): total " + totalTime + " ms, avg "
+          + (totalTime / totalPages) + " ms/page, max " + maxPageTime + " ms");
+      Assert.assertTrue(
+          "Avg page time too slow: " + (totalTime / totalPages) + " ms (max: 200 ms)",
+          (totalTime / totalPages) <= 200);
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/QueryStressTest.java]: **Suggestion**

Extract the average page time calculation to a variable before using it in the assertion message and condition.

```java
      long avgPageTime = totalTime / totalPages;
      Assert.assertTrue(
          "Avg page time too slow: " + avgPageTime + " ms (max: 200 ms)",
          avgPageTime <= 200);
```

**Rationale**: Extracting the calculation to a local variable improves readability and avoids performing the division twice.

File: src-test/src/org/openbravo/test/stress/QueryStressTest.java [hunk 21]
```diff
@@ -0,0 +1,770 @@
...
+        OBDal.getInstance().getSession().clear();
+      }
+
+      log.info("[STRESS-QUERY] Product search (" + searches.length
+          + " queries): total " + totalTime + " ms, avg "
+          + (totalTime / searches.length) + " ms/query, " + totalResults + " total results");
+      Assert.assertTrue(
+          "Avg search too slow: " + (totalTime / searches.length) + " ms (max: 500 ms)",
+          (totalTime / searches.length) <= 500);
+    } finally {
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/QueryStressTest.java:308]: **Suggestion**

Consider adding a check for 'searches.length > 0' before the division to prevent potential ArithmeticException and extract the result to a variable for clarity.

```java
      int searchCount = searches.length;
      long avgSearchTime = searchCount > 0 ? totalTime / searchCount : 0;
      Assert.assertTrue(
          "Avg search too slow: " + avgSearchTime + " ms (max: 500 ms)",
          avgSearchTime <= 500);
```

**Rationale**: Performing a division without checking if the divisor (searches.length) is zero can lead to an ArithmeticException. While unlikely in a test setup, defensive checks are recommended to avoid runtime failures.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src-test/src/org/openbravo/test/stress/GoodsReceiptStressTest.java [hunk 22]
```diff
@@ -0,0 +1,354 @@
...
+import org.openbravo.model.materialmgmt.transaction.ShipmentInOutLine;
+
+/** Stress tests for Goods Receipt creation and completion. */
+public class GoodsReceiptStressTest extends StressTestBase {
+
+  private static final String STRESS_DOC_PREFIX = "STRESS-GR-";
+  private static final String LOG_TOTAL_TIME = "[STRESS-GR] Total time: ";
+
+  private static final long MAX_MS_PER_RECEIPT_COMPLETE = 200;
+  private static final long MAX_MS_PER_LINE_CREATION = 10;
```
Comment (etendobot) [inline src-test/src/org/openbravo/test/stress/GoodsReceiptStressTest.java]: **Suggestion**

Consider adding a brief comment explaining the reason for increasing this threshold.

```java
/** Increased threshold to 200ms to avoid flakiness in performance-limited CI environments. */\n  private static final long MAX_MS_PER_RECEIPT_COMPLETE = 200;
```

**Rationale**: Thresholds in stress tests serve as performance benchmarks. Documenting the reason for an increase helps future developers understand if the performance has degraded or if the threshold was simply too strict for certain environments.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

**Review by Matias-Bernal — APPROVED** [review]
